### PR TITLE
Optimise calculate_root cache invalidation

### PR DESCRIPTION
**Pull request by expenses**
Fixes https://github.com/paritytech/smoldot/issues/629

### CODE REVIEW

File: src/trie/calculate_root.rs [hunk 1]
```diff
@@ -169,9 +169,15 @@ impl CalculationCache {
         // We invalidate the Merkle value of `node_to_invalidate` and all its ancestors.
         node_to_invalidate.user_data().merkle_value = None;
         let mut parent = node_to_invalidate.into_parent();
-        while let Some(mut p) = parent.take() {
-            p.user_data().merkle_value = None;
-            parent = p.into_parent();
+        while let Some(mut node) = parent.take() {
+            // If the node has already had it's Merkle value invalidated, then
```
Comment (tomaka) [inline src/trie/calculate_root.rs:173]:
```suggestion
            // If the node has already had its Merkle value invalidated, then
```

**Comment by tomaka** [comment]
Change looks good.
I'm maybe a bit paranoid, but I think we should try syncing the first ~100k blocks of Polkadot to make sure that this doesn't break anything.

**Comment by expenses** [comment]
I tested this `cargo run --release` (which syncs polkadot) on this branch and it syncs the first 100k blocks just fine.

**Comment by tomaka** [comment]
Thanks

**Pull request merged**